Sparse algebra support with OOP API (#760)
fortran-lang/stdlib: doc/specs/stdlib_sparse.md

```Fortran
type(CSC_sp) :: CSC
call CSC%malloc(4,5,10)
CSC%data(:)   = real([9,4,4,7,8,-1,5,8,6,-3])
CSC%row(:)    = [1,2,4,2,3,3,4,3,4,1]
CSC%colptr(:) = [1,4,6,8,10,11]
```
<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
### `ELLPACK`: ELL-pack storage format
#### Status

Experimental

#### Description
The `ELL` format stores the data in a dense matrix of $nrows \times K$ in column major order. By imposing a constant number of zeros per row $K$, this format will incure in additional zeros being stored, but it enables efficient vectorization as memory acces are carried out by constant sized strides. 

```Fortran
type(ELL_sp) :: ELL
call ELL%malloc(num_rows=4,num_cols=5,num_nz_row=3)
ELL%data(1,1:3)   = real([9,-3,0])
ELL%data(2,1:3)   = real([4,7,0])
ELL%data(3,1:3)   = real([8,-1,8])
ELL%data(4,1:3)   = real([4,5,6])

ELL%index(1,1:3) = [1,5,0]
ELL%index(2,1:3) = [1,2,0]
ELL%index(3,1:3) = [2,3,4]
ELL%index(4,1:3) = [1,3,4]
```
<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
### `SELL-C`: The Sliced ELLPACK with Constant blocks format
#### Status

Experimental

#### Description

> The Sliced ELLPACK format `SELLC` is a variation of the `ELLPACK` format. This modification reduces the storage size compared to the `ELLPACK` format but maintaining its efficient data access scheme. It can be seen as an intermediate format between `CSR` and `ELLPACK`. For more details read [here](https://arxiv.org/pdf/1307.6209v1)

Reviewer: ```suggestion
The Sliced ELLPACK format `SELLC` is a variation of the `ELLPACK` format. This modification reduces the storage size compared to the `ELLPACK` format but maintaining its efficient data access scheme. It can be seen as an intermediate format between `CSR` and `ELLPACK`. For more details read [the reference](https://arxiv.org/pdf/1307.6209v1)
```

<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
## `spmv` - Sparse Matrix-Vector product

### Status

Experimental

### Description

Provide sparse matrix-vector product kernels for the current supported sparse matrix types.

$$y=\alpha*M*x+\beta*y$$

### Syntax

`call ` [[stdlib_sparse_spmv(module):spmv(interface)]] `(matrix,vec_x,vec_y [,alpha,beta])`

### Arguments

`matrix`, `intent(in)`: Shall be a `real` or `complex` sparse type matrix.

`vec_x`, `intent(in)`: Shall be a rank-1 or rank-2 array of `real` or `complex` type array.

`vec_y`, `intent(inout)`: Shall be a rank-1 or rank-2 array of `real` or `complex` type array.

`alpha`, `intent(in)`, `optional` : Shall be a scalar value of the same type as `vec_x`. Default value `alpha=1`.

`beta`, `intent(in)`, `optional` : Shall be a scalar value of the same type as `vec_x`. Default value `beta=0`.

<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
## `sparse_conversion` - Sparse matrix to matrix conversions

### Status

Experimental

### Description

This module provides facility functions for converting between storage formats.

### Syntax

> `call ` [[stdlib_sparse_conversion(module):coo2ordered(interface)]] `(coo)`

Reviewer: I like having subroutines to maximize performance and avoid the issues with function assignment. I wonder if we should also provide 'clean' function interfaces to generate new objects? Just giving it a thought. 

They would probably pair better with overloaded operators for example, but also in other ways. For example, one could have: 

```fortran
type(COO_sp) :: mat
[...]

! Convert to CSR using interface CSR_sp and then do stuff
call my_CSR_kernel( CSR_sp(mat) , x , y, z )
```

This would require to add an interface
```Fortran
interface CSR_sp
   module procedure coo2csr_fun
end interface

elemental type(CSR_sp) function coo2csr_fun(COO) result(CSR)
    type(COO_sp), intent(in) :: COO
    call coo2csr( COO, CSR )
end function
```

### Arguments

`COO`, `intent(inout)`: Shall be any `COO` type. The same object will be returned with the arrays reallocated to the correct size after removing duplicates.

> `sort_data`, `logical(in)`, `optional`:: Boolean optional to determine whether to sort the data in the COO graph while sorting the index array, defult `.false.`.

Reviewer: ```suggestion
`sort_data`, `logical(in)`, `optional`:: Shall be an optional `logical` argument to determine whether data in the COO graph should be sorted while sorting the index array, defult `.false.`.
```

---
title: sparse
---

# The `stdlib_sparse` module

[TOC]

## Introduction

The `stdlib_sparse` module provides derived types for standard sparse matrix data structures. It also provides math kernels such as sparse matrix-vector product and conversion between matrix types.

## Sparse matrix derived types

<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
### The `sparse_type` abstract derived type
#### Status

Experimental

#### Description
The parent `sparse_type` is as an abstract derived type holding the basic common meta data needed to define a sparse matrix, as well as shared APIs. All sparse matrix flavors are extended from the `sparse_type`.

```Fortran
type, public, abstract :: sparse_type
    integer :: nrows   !! number of rows
    integer :: ncols   !> number of columns
    integer :: nnz     !> number of non-zero values
    integer :: storage !> assumed storage symmetry

> end type

Reviewer 2: Just a quick thought: it might be of interest to have two DTs with different kinds for the integers, e.g.,
```fortran
type,...... :: sparse_int32_type
 ....
 integer(int32) :: nnz
 ...
end type
type,...... :: sparse_int64_type
 ....
 integer(int64) :: nnz
 ...
end type
````

Reviewer: I'm not sure how hard it would be to accomplish this. Maybe in the current PR it would be enough to generalize the index `kind` from default `integer` to `integer(ik)`, so that later this extension will be easier. 

What you're suggesting is probably that there should be two `abstract` sparse types, one per length `kind`: 

```Fortran
type, public, abstract :: sparse_type

end type
#:for ik,it in INT_KINDS_TYPES
type, public, abstract, extends(sparse_type) :: sparse_${ik}$_type
    ${it}$ :: nrows   !! number of rows
    ${it}$ :: ncols   !> number of columns
    ${it}$ :: nnz     !> number of non-zero values
    ${it}$ :: storage !> assumed storage symmetry
end type sparse_${ik}$_type
!:endfor
```

Reviewer 2: Indeed, I had something similar to your proposition in mind. However, I agree that it should be for a following PR.

Author: I get the feeling that adding this `COO_${s1}$_type` plus then the different integer sizes would end up with declarations such as:
```Fortran
type(CSR_int32_dp_type) :: mymatrix
```
wich I find very verbose. The target design I was striving for is to have these two things being syntactically close from a hierarchical point of view:
```Fortran
real(dp), allocatable :: dense(:,:)
type(CSR_dp) :: sparse
```
Is there a possibility to find a middle ground?

Author: In https://github.com/fortran-lang/stdlib/pull/760/commits/6ae038bd54ab9927aec20bb0787ba89fa5769c17 I added a `ilp` parameter following the linalg module to enable recompiling using int64. Rightnow it is set to int32, we could add a preprocessing macro to change it.

Reviewer: > possibility to find a middle ground

I agree with you @jalvesz. @jvdp1's idea is correct, but maybe, who needs two index types when you can have the largest one already? In other words, could we just require that the index type for sparse matrices be a 64-bit one and define  

```
integer, parameter :: index_t = int64
```

To support 32-bit index types, one idea could be to just define the interface on the abstract type: 

```
type, abstract :: sparse
 ...
contains
   procedure(get_elem), deferred, abstract :: get_int64
   procedure :: get_int32 ! implemented, just calls get_int64
   generic :: get => get_int32, get_int64
end type
...
elemental real(dp) function get_int32(this,i,j)
    class(sparse), intent(in) :: this
    integer(int32), intent(in) :: i,j
    get_int32 = this%get(int(i,int64),int(j,int64))
end function
```

Author: >In other words, could we just require that the index type for sparse matrices be always a 64-bit

I'm very hesitant about this. given that with int32 one can have `nnz = 2,147,483,647` this means that, for a COO matrix of double precision values of such size, one would have 2147483647 *( 2 * 4 + 8 ) / (1024)^3 = 23 Gb of memory. Transforming the index arrays to int64 => 2147483647 *( 2 * 8 + 8 ) / (1024)^3 = 39 Gb ... And all the loops will be done by searching on larger sized indexing arrays, thus plummetting performance (do not know to which extent). Also, MKL uses int32 for the kind definition of index arrays. Shall we set a default to int64 this would make it more complicated to eventually bind against it.

Thinking about this makes me more inclined for either allowing recompilation by changing the `ilp` parameter, or indeed adding support for both sizes with separate kinds. I would in that case let the int32 with a simplified syntax and the int64 one with something like `long` added as a suffix.

Reviewer: I agree with you @jalvesz: at this stage, probably the best thing to do is just to generalize the integer kind: 

> from default `integer` to `integer(ik)`

so that the implementation already supports any integer `kind`s, but a more thoughtful decision about what should be used and how flexible should that be, is deferred.

Reviewer 2: > could we just require that the index type for sparse matrices be always a 64-bit one

The default kind should be `integer(int32)`, at least for linking them or using other libraries (like MKL Sparse BLAS).

At this stage, using `integer(ik)` is fine. There is enough WIP without this feature.

> end type

Author: I didn't see the need as the DTs only have simple integer scalars or allocatable arrays as members. My understanding is that such types are normally well destroyed by the compilers.

```Fortran
type(ELL_sp) :: ELL
call ELL%malloc(num_rows=4,num_cols=5,num_nz_row=3)
ELL%data(1,1:3)   = real([9,-3,0])
ELL%data(2,1:3)   = real([4,7,0])
ELL%data(3,1:3)   = real([8,-1,8])
ELL%data(4,1:3)   = real([4,5,6])

ELL%index(1,1:3) = [1,5,0]
ELL%index(2,1:3) = [1,2,0]
ELL%index(3,1:3) = [2,3,4]
ELL%index(4,1:3) = [1,3,4]
```
<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
### `SELL-C`: The Sliced ELLPACK with Constant blocks format
#### Status

Experimental

#### Description
The Sliced ELLPACK format `SELLC` is a variation of the `ELLPACK` format. This modification reduces the storage size compared to the `ELLPACK` format but maintaining its efficient data access scheme. It can be seen as an intermediate format between `CSR` and `ELLPACK`. For more details read [the reference](https://arxiv.org/pdf/1307.6209v1)

<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->

> ## `add`/`at` - Sparse Matrix data accessors

Author: Splitted it into two subsections, one for `add` and another for `at`

### Status

Experimental

### Description

> Type-bound procedures to enable adding or requesting data in/from a sparse matrix.

Author: It just adds values into pre-existing entries in a given sparse pattern.

### Syntax

`call matrix%add(i,j,v)` or
`call matrix%add(i(:),j(:),v(:,:))`

### Arguments

> `i`, `intent(in)`: Shall be an integer value or rank-1 array.

Author: done in https://github.com/fortran-lang/stdlib/pull/760/commits/dde88a732d67a0b0bd4425d7734ffeccc53e28d0

`j`, `intent(in)`: Shall be an integer value or rank-1 array.
`v`, `intent(in)`: Shall be a `real` or `complex` value or rank-2 array. The type shall be in accordance to the declared sparse matrix object.

> ### Syntax

Author: done in https://github.com/fortran-lang/stdlib/pull/760/commits/dde88a732d67a0b0bd4425d7734ffeccc53e28d0

> 

Author: `at` returns a value from the sparse pattern. Just like from a dense matrix one would request a value by `val=dense(i,j)`, here we propose `val = sparse%at(i,j)` to mimic a dense notation.

`v = matrix%at(i,j)`

### Arguments

`i`, `intent(in)` : Shall be an integer value.
`j`, `intent(in)` : Shall be an integer value.

> `v`, `result`     : Shall be a `real` or `complex` value in accordance to the declared sparse matrix object. If the `ij` tuple is within the sparse pattern, `v` returns the value in the data buffer. If the `ij` tuple is outside the sparse pattern, `v` returns `0`. If the `ij` tuple is outside the matrix pattern `(nrows,ncols)`, `v` returns `NaN`.

Author: done in https://github.com/fortran-lang/stdlib/pull/760/commits/dde88a732d67a0b0bd4425d7734ffeccc53e28d0

## Example
```fortran
{!example/linalg/example_sparse_data_accessors.f90!}
```

<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
## `spmv` - Sparse Matrix-Vector product

### Status

Experimental

### Description

Provide sparse matrix-vector product kernels for the current supported sparse matrix types.

$$y=\alpha*M*x+\beta*y$$

### Syntax

`call ` [[stdlib_sparse_spmv(module):spmv(interface)]] `(matrix,vec_x,vec_y [,alpha,beta])`

### Arguments

`matrix`, `intent(in)`: Shall be a `real` or `complex` sparse type matrix.

`vec_x`, `intent(in)`: Shall be a rank-1 or rank-2 array of `real` or `complex` type array.

`vec_y`, `intent(inout)`: Shall be a rank-1 or rank-2 array of `real` or `complex` type array.

`alpha`, `intent(in)`, `optional` : Shall be a scalar value of the same type as `vec_x`. Default value `alpha=1`.

`beta`, `intent(in)`, `optional` : Shall be a scalar value of the same type as `vec_x`. Default value `beta=0`.

<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
## `sparse_conversion` - Sparse matrix to matrix conversions

### Status

Experimental

### Description

> This module provides facility functions for converting between storage formats.

Author: There are several conversion procedures within the module

`row`, `integer(in)`:: rows index array.

`col`, `integer(in)`:: columns index array.

`data`, `real/complex(in)`, `optional`:: data array.

`nrows`, `integer(in)`, `optional`:: number of rows, if not given it will be computed from the `row` array.

`ncols`, `integer(in)`, `optional`:: number of columns, if not given it will be computed from the `col` array.

`num_nz_rows`, `integer(in)`, `optional`:: number of non zeros per row, only valid in the case of an `ELL` matrix, by default it will computed from the largest row.

`chunk`, `integer(in)`, `optional`:: chunk size, only valid in the case of a `SELLC` matrix, by default it will be taken from the `SELLC` default attribute chunk size.

## Example
```fortran
{!example/linalg/example_sparse_from_ijv.f90!}
```

### Syntax

`call ` [[stdlib_sparse_conversion(module):dense2coo(interface)]] `(dense,coo)`

### Arguments

`dense`, `intent(in)`: Shall be a rank-2 array of `real` or `complex` type.

> `coo`, `intent(inout)`: Shall be a `COO` type of `real` or `complex` type.

Author: change to `intent(out)` in https://github.com/fortran-lang/stdlib/pull/760/commits/dde88a732d67a0b0bd4425d7734ffeccc53e28d0 for all conversion procedures except for coo2ordered.

`COO`, `intent(inout)`: Shall be any `COO` type. The same object will be returned with the arrays reallocated to the correct size after removing duplicates.

> `sort_data`, `logical(in)`, `optional`:: Shall be an optional `logical` argument to determine whether data in the COO graph should be sorted while sorting the index array, defult `.false.`.

Reviewer 3: ```suggestion
`sort_data`, `logical(in)`, `optional`:: Shall be an optional `logical` argument to determine whether data in the COO graph should be sorted while sorting the index array, default `.false.`.
```
